src/main/scala/format/pgn/Parser.scala

```diff
@@ -98,296 +63,200 @@ object Parser {
+    // B@g5
+    val drop: P[Drop] = ((role <* P.char('@')) ~ dest).map { case (role, pos) => Drop(role, pos) }
 
-    val checkmate = ("#" | "++") ^^^ true | success(false)
+    val pawnDrop: P[Drop] = (P.char('@') *> dest).map(Drop(Pawn, _))
 
-    val role = mapParser(Role.allByPgn, "role") | success(Pawn)
+    // Bac3 Baxc3 B2c3 B2xc3 Ba2xc3
+    val disambiguated: P[Std] = (role ~ file.? ~ rank.? ~ x ~ dest).map { case ((((ro, fi), ra), ca), de) =>
+      Std(dest = de, role = ro, capture = ca, file = fi, rank = ra)
+    }
 
-    val file = mapParser(fileMap, "file")
+    // d7d5
+    val disambiguatedPawn: P[Std] = (((file.? ~ rank.?) ~ x).with1 ~ dest).map { case (((fi, ra), ca), de) =>
+      Std(dest = de, role = Pawn, capture = ca, file = fi, rank = ra)
+    }
 
-    val rank = mapParser(rankMap, "rank")
+    val suffixes: P0[Suffixes] = (promotion.? ~ checkmate ~ check ~ glyphs).map { case (((p, cm), c), g) =>
+      Suffixes(c, cm, p, g)
+    }
 
-    val promotable = Role.allPromotableByPgn mapKeys (_.toUpper)
+    val castle: P[San] = (qCastle | kCastle).map(Castle(_))
 
-    val promotion = ("=" ?) ~> mapParser(promotable, "promotion")
+    val standard: P[San] =
+      disambiguatedPawn.backtrack | pawn.backtrack | disambiguated.backtrack | ambigous.backtrack | drop.backtrack | pawnDrop.backtrack
 
-    val dest = mapParser(Pos.allKeys, "dest")
+    val move: P[San] = (castle | standard).withContext("Invalid chess move")
+    val moveWithSuffix: P[San] = (move ~ suffixes <* whitespaces)
+      .map { case (std, suf) =>
+        std withSuffixes suf
+      }
 
-    def exists(c: String): Parser[Boolean] = c ^^^ true | success(false)
+    def apply(str: String): Validated[String, San] =
+      moveWithSuffix.parse(str) match {
+        case Right((_, san)) => valid(san)
+        case Left(err)       => invalid(showExpectations("Cannot parse move", str, err))
+      }
 
-    def mapParser[A, B](pairs: Iterable[(A, B)], name: String): Parser[B] =
-      pairs.foldLeft(failure(name + " not found"): Parser[B]) { case (acc, (a, b)) =>
-        a.toString ^^^ b | acc
+    def mapParser[A, B](pairs: Iterable[(A, B)], name: String): P[B] =
+      pairs.foldLeft(P.failWith(name + " not found"): P[B]) { case (acc, (a, b)) =>
+        P.string(a.toString).as(b) | acc
       }
+
   }
 
-  object TagParser extends RegexParsers with Logging {
+  object TagParser {
+
+    val tagName: P[String]      = R.alpha.rep.string
+    val escapeDquote: P[String] = (P.char('\\') ~ R.dquote).as("\"")
+    val valueChar: P[String]    = escapeDquote | P.charWhere(_ != '"').string
+    val tagValue: P[String]     = valueChar.rep0.map(_.mkString).with1.surroundedBy(R.dquote)
+    val tagContent: P[Tag]      = ((tagName <* R.wsp.rep) ~ tagValue).map(p => Tag(p._1, p._2))
+    val tag: P[Tag]             = tagContent.between(P.char('['), P.char(']')) <* whitespace.rep0
+    val tags: P[Tags]           = tag.backtrack.rep.map(tags => Tags(tags.toList))
```

Comment:
I wonder if you need the `backtrack` here since tags are between `[` which I guess means if you fail consuming something you have a real failure. Also you could do `tag.repAs[List[Tag]].map(Tags(_))` and it would be modestly more efficient.

Reply:
The old implementation supports some weird/incorrect tags at the end, e.g 
```
[abc "abc"]
[def]
```
is accepted as correct tags. So that's why I need backtrack to a void making it a real failure (which I prefer and what we should do anyway).